zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
+    handler::F
 end
 
-function __zcm_handler_wrapper(rbuf::Ptr{Native.RecvBuf}, channel::Cstring, usr_::Ptr{Void})
-    usr = unsafe_pointer_to_objref(usr_)::__zcm_handler;
-    usr.handler(RecvBuf(rbuf), unsafe_string(channel), usr.usr);
-    return nothing;
-end
-
-type Sub
-    sub::Ptr{Native.Sub};
-
-    # References to protect our callbacks from Julia's garbage collector
-    handler_jl::__zcm_handler;
-    handler_c ::Ptr{Void};
-    uv_handler::Ptr{Void};
-    uv_wrapper::Ptr{Native.UvSub};
-
-    Sub(sub::Ptr{Native.Sub}, handler_jl::__zcm_handler, handler_c::Ptr{Void},
-        uv_handler::Ptr{Void}, uv_wrapper::Ptr{Native.UvSub}) =
-        new(sub, handler_jl, handler_c, uv_handler, uv_wrapper);
+"""
+The Subscription type contains the SubscriptionOptions and also all of the
+various C pointers to the libuv handlers. 
+"""
+immutable Subscription{T <: SubscriptionOptions}
+    options::T
+    c_handler::Ptr{Void}
+    uv_wrapper::Ptr{Native.UvSub}
+    uv_handler::Ptr{Void}
+    native_sub::Ptr{Native.Sub}
 end
 
 type Zcm
-    zcm::Ptr{Native.Zcm};
-
-    # http://docs.julialang.org/en/stable/manual/calling-c-and-fortran-code/
-    # http://julialang.org/blog/2013/05/callback
-
-    # RRR: make prints throughout this file optional. Note that printing in a finalizer is not
-    #      ok because it can cause errors on exit if stdout has already been cleaned up
-    function Zcm(url::AbstractString)
-        instance = new();
-        instance.zcm = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), url);
-
-        # user can force cleanup of their instance by calling `finalize(zcm)`
-        finalizer(instance, function(zcm::Zcm)
-                                if (zcm.zcm != C_NULL)
-                                    ccall(("zcm_destroy", "libzcm"), Void,
-                                          (Ptr{Native.Zcm},), zcm.zcm);
-                                    zcm.zcm = C_NULL;
-                                end
-                            end);
-
-        return instance;
+    zcm::Ptr{Native.Zcm}
+    provider::String
+    subscriptions::Vector{Subscription}
+
+    function Zcm(provider::AbstractString = "")
+        pointer = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), provider);
+        instance = new(pointer, provider, Subscription[])
+        finalizer(instance, close)
+        return instance
     end
+end
 
-    function Zcm() return Zcm(""); end
+function close(zcm::Zcm)
+    if zcm.zcm != C_NULL
+        ccall(("zcm_destroy", "libzcm"), Void,
+              (Ptr{Native.Zcm},), zcm)
+        zcm.zcm = C_NULL
+    end
 end
-export Zcm;
 
-function encode end
-function decode end
+# Defines the conversion when we pass this to a C function expecting a pointer
+Base.unsafe_convert(::Type{Ptr{Native.Zcm}}, zcm::Zcm) = zcm.zcm
 
-function good(zcm::Zcm)
-    return (zcm.zcm != C_NULL) && (errno(zcm) == 0);
-end
-export good;
+good(zcm::Zcm) = (zcm.zcm != C_NULL) && (errno(zcm) == 0)
```

Comment:
Not a fan of this syntax. Is there an issue with the more verbose multi-line function syntax?

Reply:
No, they're equivalent. This is just more compact.

Reply:
It's all the same to me. I'll switch back to the multi-line syntax.

Reply:
resolved in https://github.com/ZeroCM/zcm/pull/165/commits/2973539cb34ef53f0dc52a140bb801cdd83405ee